### PR TITLE
gMSA support on Linux - capability advertisement

### PR DESCRIPTION
**Pull request by saikiranakula-amzn**
gMSA support on Linux - capability advertisement
*changes to ecs-init to establish communication of agent with credentials-fetcher daemon
*changes to agent to advertise gMSA capability on Linux

### Summary
Agent to advertise gMSA support on Linux

### Implementation details
*changes to ecs-init to establish communication of agent with credentials-fetcher daemon
*changes to agent to advertise gMSA capability on Linux

### Testing
Added unit test to validate the funtionality
Note for external contributors:
`make test` and `make run-integ-tests` can run in a Linux development
environment like your laptop.  `go test -timeout=30s ./agent/...` and
`.\scripts\run-integ.tests.ps1` can run in a Windows development environment
like your laptop.  Please ensure unit and integration tests pass (on at least
one platform) before opening the pull request.
Once you open the pull request, there will be 14 automatic test checks on the bottom
of the pull request, please make sure they all pass before you merge it. You can
use `bot/test` label to rerun the automatic tests multiple times.
-->

New tests cover the changes: 

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/config/parse.go [hunk 1]
```diff
@@ -27,6 +27,13 @@ import (
 	cnitypes "github.com/containernetworking/cni/pkg/types"
 )
 
+const (
+	// envSkipDomainJoinCheck is an environment setting that can be used to skip
+	// domain join check validation. This is useful for integration and
+	// functional-tests but should not be set for any non-test use-case.
+	envSkipDomainJoinCheck = "ZZZ_SKIP_DOMAIN_JOIN_CHECK_NOT_SUPPORTED_IN_PRODUCTION"
```
Comment (fierlion) [inline agent/config/parse.go:34]: nit do we need the `ZZZ`?

File: ecs-init/config/common.go [hunk 2]
```diff
@@ -212,6 +218,23 @@ func DockerUnixSocket() (string, bool) {
...
+// CredentialsFetcherUnixSocket returns the credentials fetcher daemon socket endpoint and whether it reads from CredentialsFetcherEnvVar
+func CredentialsFetcherUnixSocket() (string, bool) {
+	if credentialsFetcherHost := os.Getenv(CredentialsFetcherHostEnvVar); strings.HasPrefix(credentialsFetcherHost, UnixSocketPrefix) {
+		return strings.TrimPrefix(credentialsFetcherHost, UnixSocketPrefix), true
+	}
+
+	return "/var/credentials-fetcher/socket/credentials_fetcher.sock", false
+}
+
+// HostCredentialsFetcherPath() returns the daemon socket lacation
```
Comment (fierlion) [inline ecs-init/config/common.go:230]: nit `location`

File: ecs-init/config/common.go [hunk 3]
```diff
@@ -212,6 +218,23 @@ func DockerUnixSocket() (string, bool) {
...
+func HostCredentialsFetcherPath() string {
```
Comment (fierlion) [inline ecs-init/config/common.go:231]: should this also return (string, bool) rather than "" as the missing indicator?

File: ecs-init/docker/docker.go [hunk 4]
```diff
@@ -440,6 +463,32 @@ func (c *client) getHostConfig(envVarsFromFiles map[string]string) *godocker.Hos
...
 
+// getCredentialsFetcherSocketBind returns the bind for CredentialsFetcher socket and also whether the bind exists.
+// Value for the bind is as follow:
+// CREDENTIALS_FETCHER_HOST (as in os.Getenv) set: source CREDENTIALS_FETCHER_HOST (as in os.Getenv, trim unix:// prefix),
+// dest CREDENTIALS_FETCHER_HOST (as in /etc/ecs/ecs.config, trim unix:// prefix)
+func getCredentialsFetcherSocketBind(envVarsFromFiles map[string]string) (string, bool) {
+	CredentialsFetcherEndpointAgent := defaultCredentialsFetcherSocketPath
+	CredentialsFetcherUnixSocketSourcePath, fromEnv := config.CredentialsFetcherUnixSocket()
+	if fromEnv {
+		if credentialsFetcherEndpointConfig, ok := envVarsFromFiles[config.CredentialsFetcherHostEnvVar]; ok && strings.HasPrefix(credentialsFetcherEndpointConfig, config.UnixSocketPrefix) {
```
Comment (fierlion) [inline ecs-init/docker/docker.go:474]: nit: any way to make this line more readable?  There's a lot going on here.

File: agent/config/parse.go (hunk 1 above)
Comment (saikiranakula-amzn) [reply agent/config/parse.go:34]: this flag is used by windows gMSA as well, It is possible i might break stuff if I change it. I am reusing existing one

File: agent/config/parse_linux.go [hunk 5]
```diff
@@ -18,10 +18,43 @@ package config
...
+	"github.com/cihub/seelog"
 )
 
 func parseGMSACapability() bool {
+	envStatus := utils.ParseBool(os.Getenv("ECS_GMSA_SUPPORTED"), true)
+	if envStatus {
+		// Check if domain join check override is present
+		skipDomainJoinCheck := utils.ParseBool(os.Getenv(envSkipDomainJoinCheck), false)
+		if skipDomainJoinCheck {
+			seelog.Errorf("Skipping domain join validation based on environment override")
```
Comment (singholt) [inline agent/config/parse_linux.go:34]: `seelog.Debug`?

File: agent/config/parse_linux.go [hunk 6]
```diff
@@ -18,10 +18,43 @@ package config
...
+			return true
+		}
+
+		// check if the credentials fetcher socket is created and exists
+		// this env variable is set in ecs-init module
+		if credentialsfetcherHostDir := os.Getenv("CREDENTIALS_FETCHER_HOST_DIR"); credentialsfetcherHostDir != "" {
+			_, err := os.Stat(credentialsfetcherHostDir)
+			if err != nil {
+				if os.IsNotExist(err) {
+					seelog.Errorf("credentials fetcher socket doesn't exist to support linux gmsa")
```
Comment (singholt) [inline agent/config/parse_linux.go:44]: `seelog.Error`?

File: ecs-init/docker/docker.go (hunk 4 above)
Comment (saikiranakula-amzn) [reply ecs-init/docker/docker.go]: made the change

File: ecs-init/config/common.go (hunk 3 above)
Comment (saikiranakula-amzn) [reply ecs-init/config/common.go:231]: following the pattern of other Host methods

File: agent/config/config_unix.go [hunk 7]
```diff
@@ -100,7 +100,7 @@ func DefaultConfig() Config {
 		PollingMetricsWaitDuration:          DefaultPollingMetricsWaitDuration,
 		NvidiaRuntime:                       DefaultNvidiaRuntime,
 		CgroupCPUPeriod:                     defaultCgroupCPUPeriod,
-		GMSACapable:                         false,
+		GMSACapable:                         true,
```
[the remark this answers is not in the document]
Comment (saikiranakula-amzn) [reply agent/config/config_unix.go]: made change to read the env variable from the config

File: agent/config/parse_linux.go [hunk 8]
```diff
@@ -18,10 +18,43 @@ package config
...
+					return false
+				}
+			}
+			// returns true if the container instance is domain joined
+			// this env variable is set in ecs-init module
+			isDomainJoined := utils.ParseBool(os.Getenv("ECS_DOMAIN_JOINED_LINUX_INSTANCE"), false)
+
+			if !isDomainJoined {
+				seelog.Errorf("Unable to determine valid domain join")
```
Comment (singholt) [inline agent/config/parse_linux.go:53]: `seelog.Error`?

File: agent/config/parse_linux_test.go [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+import (
+	"os"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestParseGMSACapabilitySupported(t *testing.T) {
+	os.Setenv("ECS_GMSA_SUPPORTED", "True")
```
Comment (singholt) [inline agent/config/parse_linux_test.go:27]: nit: something I learnt recently - you can use `t.Setenv` to set the env variable. it would cleanup as part of the test cleanup workflow without explicit defer statements. 

https://pkg.go.dev/testing#B.Setenv

applicable to multiple tests in this PR

File: agent/config/config_unix.go [hunk 10]
```diff
@@ -100,7 +100,7 @@ func DefaultConfig() Config {
 		PollingMetricsWaitDuration:          DefaultPollingMetricsWaitDuration,
 		NvidiaRuntime:                       DefaultNvidiaRuntime,
 		CgroupCPUPeriod:                     defaultCgroupCPUPeriod,
-		GMSACapable:                         false,
+		GMSACapable:                         isGMSACapable(),
```
Comment (singholt) [inline agent/config/config_unix.go:103]: shouldn't this call `parseGMSACapability()` which is in `parse_linux.go`

File: agent/config/config_unix.go [hunk 11]
```diff
@@ -135,3 +135,7 @@ func (cfg *Config) platformString() string {
 func getConfigFileName() (string, error) {
 	return utils.DefaultIfBlank(os.Getenv("ECS_AGENT_CONFIG_FILE_PATH"), defaultConfigFileName), nil
 }
+
+func isGMSACapable() bool {
```
Comment (singholt) [inline agent/config/config_unix.go:139]: I dont think this is needed.

also `parseGMSACapability()` is checking this env var and setting default to true whereas this is setting to false.
can we remove this?

File: ecs-init/config/common.go [hunk 12]
```diff
@@ -212,6 +218,23 @@ func DockerUnixSocket() (string, bool) {
 	return "/var/run", false
 }
 
+// CredentialsFetcherUnixSocket returns the credentials fetcher daemon socket endpoint and whether it reads from CredentialsFetcherEnvVar
+func CredentialsFetcherUnixSocket() (string, bool) {
```
Comment (singholt) [inline ecs-init/config/common.go:222]: does this need to be exported?

File: ecs-init/config/common_test.go [hunk 13]
```diff
@@ -253,3 +253,30 @@ func TestAgentRunningInExternal(t *testing.T) {
 	defer os.Unsetenv(ExternalEnvVar)
 	assert.True(t, RunningInExternal())
 }
+
+func TestCredentialsFetcherUnixSocketWithoutCredentialsFetcherHost(t *testing.T) {
+	// Make sure that the env variable is not set
+	os.Unsetenv("CREDENTIALS_FETCHER_HOST")
```
Comment (singholt) [inline ecs-init/config/common_test.go:259]: I dont think this is needed, if you've not set it, it wont be.

also using t.Setenv might help with unsurety

File: agent/config/parse_linux.go (hunk 6 above)
Comment (sparrc) [inline agent/config/parse_linux.go:44]: Make this error message more explicit, something like

```go
fmt.Sprintf"CREDENTIALS_FETCHER_HOST_DIR not found, was looking for %s", credentialsfetcherHostDir)
```

File: ecs-init/config/common.go [hunk 14]
```diff
@@ -212,6 +218,23 @@ func DockerUnixSocket() (string, bool) {
...
+	if credentialsFetcherHost := os.Getenv(CredentialsFetcherHostEnvVar); strings.HasPrefix(credentialsFetcherHost, UnixSocketPrefix) {
+		return strings.TrimPrefix(credentialsFetcherHost, UnixSocketPrefix), true
+	}
+
+	return "/var/credentials-fetcher/socket/credentials_fetcher.sock", false
```
Comment (singholt) [inline ecs-init/config/common.go:227]: can we use `defaultCredentialsFetcherSocketPath` here?

File: agent/config/parse_linux.go (hunk 8 above)
Comment (sparrc) [inline agent/config/parse_linux.go:53]: ditto above comment, make this more explicit, something like
```go
"gMSA on linux requires domain joined instance. Did not find expected env var ECS_DOMAIN_JOINED_LINUX_INSTANCE=true"
```

File: agent/config/parse_linux.go [hunk 15]
```diff
@@ -18,10 +18,43 @@ package config
...
+		// Check if domain join check override is present
+		skipDomainJoinCheck := utils.ParseBool(os.Getenv(envSkipDomainJoinCheck), false)
+		if skipDomainJoinCheck {
+			seelog.Errorf("Skipping domain join validation based on environment override")
+			return true
+		}
+
+		// check if the credentials fetcher socket is created and exists
+		// this env variable is set in ecs-init module
+		if credentialsfetcherHostDir := os.Getenv("CREDENTIALS_FETCHER_HOST_DIR"); credentialsfetcherHostDir != "" {
```
Comment (sparrc) [inline agent/config/parse_linux.go:40]: Why aren't you using `common.CredentialsFetcherUnixSocket()` here?

File: ecs-init/config/common.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (sparrc) [reply ecs-init/config/common.go:227]: Also this string should be a constant for doc purposes.

File: ecs-init/config/common.go (hunk 3 above)
Comment (singholt) [reply ecs-init/config/common.go:231]: I also agree with Ray. returning string, bool seems much safer than checking for empty strings.

File: ecs-init/docker/docker.go [hunk 16]
```diff
@@ -49,6 +50,10 @@ const (
 	// in case /var/run/docker.sock is deleted and recreated outside the container
 	defaultDockerEndpoint   = "/var/run"
 	defaultDockerSocketPath = "/var/run/docker.sock"
+	// this socket is exposed by credentials-fetcher (daemon for gMSA support on Linux)
+	// defaultCredentialsFetcherSocketPath is set to /var/credentials-fetcher/socket/credentials_fetcher.sock
+	// in case path is not passed in the env variable
+	defaultCredentialsFetcherSocketPath = "/var/credentials-fetcher/socket/credentials_fetcher.sock"
```
Comment (sparrc) [inline ecs-init/docker/docker.go:56]: This appears to be defined in multiple places. You should make it an exported constant somewhere it can be reused.

File: ecs-init/docker/docker.go [hunk 17]
```diff
@@ -440,6 +463,33 @@ func (c *client) getHostConfig(envVarsFromFiles map[string]string) *godocker.Hos
 	return createHostConfig(binds)
 }
 
+// getCredentialsFetcherSocketBind returns the bind for CredentialsFetcher socket and also whether the bind exists.
+// Value for the bind is as follow:
+// CREDENTIALS_FETCHER_HOST (as in os.Getenv) set: source CREDENTIALS_FETCHER_HOST (as in os.Getenv, trim unix:// prefix),
```
Comment (sparrc) [inline ecs-init/docker/docker.go:468]: I am having trouble understanding this comment. What does this mean? Can you rephrase it?

File: ecs-init/docker/docker.go [hunk 18]
```diff
@@ -440,6 +463,33 @@ func (c *client) getHostConfig(envVarsFromFiles map[string]string) *godocker.Hos
...
+			CredentialsFetcherEndpointAgent = strings.TrimPrefix(credentialsFetcherEndpointConfig, config.UnixSocketPrefix)
+		} else {
+			CredentialsFetcherEndpointAgent = defaultCredentialsFetcherSocketPath
+		}
+	}
+
+	// check whether the path to the mount exists
+	_, err := os.Stat(CredentialsFetcherUnixSocketSourcePath)
+	if err != nil {
+		if os.IsNotExist(err) {
```
Comment (sparrc) [inline ecs-init/docker/docker.go:485]: Why is this being checked in multiple places?

Comment (saikiranakula-amzn) [reply ecs-init/docker/docker.go:485]: one check on the init and other on the agent as well, to check the socket is properly bind mount

File: ecs-init/config/common.go [hunk 19]
```diff
@@ -93,6 +93,12 @@ const (
 
 	// DefaultRegionEnvVar is the environment variable for specifying the default AWS region to use.
 	DefaultRegionEnvVar = "AWS_DEFAULT_REGION"
+
+	// ECSGMSASupportEnvVar indicates that the gMSA is supported
+	ECSGMSASupportEnvVar = "ECS_GMSA_SUPPORTED"
```
Comment (singholt) [inline ecs-init/config/common.go:98]: it seems like we're introducing 3 new env vars:
`ECS_GMSA_SUPPORTED`
`CREDENTIALS_FETCHER_HOST`
`ECS_DOMAIN_JOINED_LINUX_INSTANCE`

- can we update the README to tell which one is configurable vs which one isnt? i imagine `ECS_DOMAIN_JOINED_LINUX_INSTANCE` shouldn't be configurable?
- can you initialize the `ECS_DOMAIN_JOINED_LINUX_INSTANCE` in this list as well and reuse the var?

File: ecs-init/docker/docker.go [hunk 20]
```diff
@@ -440,6 +463,33 @@ func (c *client) getHostConfig(envVarsFromFiles map[string]string) *godocker.Hos
...
 }
 
+// getCredentialsFetcherSocketBind returns the bind for CredentialsFetcher socket and also whether the bind exists.
+// Value for the bind is as follow:
+// CREDENTIALS_FETCHER_HOST (as in os.Getenv) set: source CREDENTIALS_FETCHER_HOST (as in os.Getenv, trim unix:// prefix),
+// dest CREDENTIALS_FETCHER_HOST (as in /etc/ecs/ecs.config, trim unix:// prefix)
+func getCredentialsFetcherSocketBind(envVarsFromFiles map[string]string) (string, bool) {
+	CredentialsFetcherEndpointAgent := defaultCredentialsFetcherSocketPath
+	CredentialsFetcherUnixSocketSourcePath, fromEnv := config.CredentialsFetcherUnixSocket()
+	if fromEnv {
```
Comment (sparrc) [inline ecs-init/docker/docker.go:473]: why are we doing special parsing only in the case where the host socket path comes from env var?

File: ecs-init/docker/docker.go [hunk 21]
```diff
@@ -440,6 +463,33 @@ func (c *client) getHostConfig(envVarsFromFiles map[string]string) *godocker.Hos
 	return createHostConfig(binds)
 }
 
+// getCredentialsFetcherSocketBind returns the bind for CredentialsFetcher socket and also whether the bind exists.
+// Value for the bind is as follow:
+// CREDENTIALS_FETCHER_HOST (as in os.Getenv) set: source CREDENTIALS_FETCHER_HOST (as in os.Getenv, trim unix:// prefix),
+// dest CREDENTIALS_FETCHER_HOST (as in /etc/ecs/ecs.config, trim unix:// prefix)
+func getCredentialsFetcherSocketBind(envVarsFromFiles map[string]string) (string, bool) {
+	CredentialsFetcherEndpointAgent := defaultCredentialsFetcherSocketPath
+	CredentialsFetcherUnixSocketSourcePath, fromEnv := config.CredentialsFetcherUnixSocket()
```
Comment (sparrc) [inline ecs-init/docker/docker.go:472]: nit: rename this CredentialsFetcherUnixSocketHostPath

File: ecs-init/docker/docker.go [hunk 22]
```diff
@@ -440,6 +463,33 @@ func (c *client) getHostConfig(envVarsFromFiles map[string]string) *godocker.Hos
 	return createHostConfig(binds)
 }
 
+// getCredentialsFetcherSocketBind returns the bind for CredentialsFetcher socket and also whether the bind exists.
+// Value for the bind is as follow:
+// CREDENTIALS_FETCHER_HOST (as in os.Getenv) set: source CREDENTIALS_FETCHER_HOST (as in os.Getenv, trim unix:// prefix),
+// dest CREDENTIALS_FETCHER_HOST (as in /etc/ecs/ecs.config, trim unix:// prefix)
+func getCredentialsFetcherSocketBind(envVarsFromFiles map[string]string) (string, bool) {
+	CredentialsFetcherEndpointAgent := defaultCredentialsFetcherSocketPath
```
Comment (sparrc) [inline ecs-init/docker/docker.go:471]: nit: rename this CredentialsFetcherUnixSocketAgentPath

File: ecs-init/docker/docker.go (hunk 20 above)
Comment (saikiranakula-amzn) [reply ecs-init/docker/docker.go:473]: yes, this it if the variable comes from env

File: agent/config/parse_linux.go (hunk 15 above)
Comment (saikiranakula-amzn) [reply agent/config/parse_linux.go:40]: common.go is in init module, this method is in agent.go

File: agent/config/parse_linux.go (hunk 5 above)
Comment (saikiranakula-amzn) [reply agent/config/parse_linux.go:34]: made it as Info as per cam's earlier comment

File: ecs-init/config/common.go (hunk 12 above)
Comment (saikiranakula-amzn) [reply ecs-init/config/common.go:222]: yes it is used in docker package

File: ecs-init/docker/docker.go (hunk 17 above)
Comment (saikiranakula-amzn) [reply ecs-init/docker/docker.go:468]: did rephrase

File: agent/config/parse_linux.go (hunk 6 above)
Comment (saikiranakula-amzn) [reply agent/config/parse_linux.go]: made the change

File: ecs-init/docker/docker.go (hunk 16 above)
Comment (saikiranakula-amzn) [reply ecs-init/docker/docker.go]: only defined in common.go and reusing it

File: agent/config/parse_linux_test.go (hunk 9 above)
Comment (saikiranakula-amzn) [reply agent/config/parse_linux_test.go]: made changes to use t.setenv

File: agent/config/config_unix.go (hunk 10 above)
Comment (saikiranakula-amzn) [reply agent/config/config_unix.go]: good point, made change accordingly

File: ecs-init/config/common.go (hunk 19 above)
Comment (saikiranakula-amzn) [reply ecs-init/config/common.go:98]: ECS_GMSA_SUPPORTED is existing variable
CREDENTIALS_FETCHER_HOST - will add to readme
ECS_DOMAIN_JOINED_LINUX_INSTANCE this is not configurable, ecs-init will set this variable

File: ecs-init/config/common_test.go (hunk 13 above)
Comment (saikiranakula-amzn) [reply ecs-init/config/common_test.go:259]: making sure this variable is not set anywhere else

File: ecs-init/config/common.go (hunk 3 above)
Comment (saikiranakula-amzn) [reply ecs-init/config/common.go]: updated accordingly

File: ecs-init/config/common.go (hunk 19 above)
Comment (saikiranakula-amzn) [reply ecs-init/config/common.go:98]: added the variable CREDENTIALS_FETCHER_HOST  to readme

File: README.md [hunk 23]
```diff
@@ -203,6 +203,7 @@ additional details on each available environment variable.
...
 | `HTTP_PROXY` | `10.0.0.131:3128` | The hostname (or IP address) and port number of an HTTP proxy to use for the Amazon ECS agent to connect to the internet. For example, this proxy will be used if your container instances do not have external network access through an Amazon VPC internet gateway or NAT gateway or instance. If this variable is set, you must also set the NO_PROXY variable to filter Amazon EC2 instance metadata and Docker daemon traffic from the proxy. | `null` | `null` |
 | `NO_PROXY` | <For Linux: 169.254.169.254,169.254.170.2,/var/run/docker.sock &#124; For Windows: 169.254.169.254,169.254.170.2,\\.\pipe\docker_engine> | The HTTP traffic that should not be forwarded to the specified HTTP_PROXY. You must specify 169.254.169.254,/var/run/docker.sock to filter Amazon EC2 instance metadata and Docker daemon traffic from the proxy. | `null` | `null` |
+| `CREDENTIALS_FETCHER_HOST` | `unix:///var/credentials-fetcher/socket/credentials_fetcher.sock` | Used to create a connection to the Credentials-fetcher daemon; to support. gMSA on Linux. | `unix:///var/credentials-fetcher/socket/credentials_fetcher.sock` | Not Applicable |
```
Comment (sparrc) [inline README.md:206]: nit: doc reword request:

```
Used to create a connection to the [credentials-fetcher daemon](https://github.com/aws/credentials-fetcher); to support gMSA on Linux. The default is fine for most users, only needs to be modified if user is configuring a custom credentials-fetcher socket path, ie, [CF_UNIX_DOMAIN_SOCKET_DIR](https://github.com/aws/credentials-fetcher#default-environment-variables).
```

rendered:

Used to create a connection to the [credentials-fetcher daemon](https://github.com/aws/credentials-fetcher); to support gMSA on Linux. The default is fine for most users, only needs to be modified if user is configuring a custom credentials-fetcher socket path, ie, [CF_UNIX_DOMAIN_SOCKET_DIR](https://github.com/aws/credentials-fetcher#default-environment-variables).

File: agent/config/parse_linux.go (hunk 15 above)
Comment (saikiranakula-amzn) [reply agent/config/parse_linux.go:40]: as from discussion common.go resides in init we can't use it

File: ecs-init/config/common.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (saikiranakula-amzn) [reply ecs-init/config/common.go]: made changes accordingly to use constant

File: README.md (hunk 23 above)
Comment (saikiranakula-amzn) [reply README.md]: updated the comment accordingly

**Review by sparrc — APPROVED** [review]

File: ecs-init/docker/docker.go [hunk 24]
```diff
@@ -541,3 +576,32 @@ func (c *client) StopAgent() error {
 	}
 	return err
 }
+
+// isDomainJoined is used to validate if container instance is part of a valid active directory.
+func isDomainJoined() bool {
+	realmPath, err := exec.LookPath("realm")
+	if err != nil {
+		log.Errorf("realmd is not installed on the instance")
```
Comment (singholt) [inline ecs-init/docker/docker.go:584]: `log.Error` or log the `err`?

File: ecs-init/docker/docker.go [hunk 25]
```diff
@@ -541,3 +576,32 @@ func (c *client) StopAgent() error {
...
+		return false
+	}
+
+	if len(realmPath) == 0 {
+		log.Errorf("could not path of realm on the instance")
```
Comment (singholt) [inline ecs-init/docker/docker.go:589]: `log.Error`

File: ecs-init/docker/docker.go [hunk 26]
```diff
@@ -541,3 +576,32 @@ func (c *client) StopAgent() error {
...
+		return false
+	}
+
+	realmCommmand := realmPath + " list"
+	realmCmdOutput, err := exec.Command("bash", "-c", realmCommmand).Output()
+	if err != nil {
+		log.Errorf("failed to read realm info")
```
Comment (singholt) [inline ecs-init/docker/docker.go:596]: `log.Error` or log the `err`?

File: ecs-init/docker/docker.go [hunk 27]
```diff
@@ -541,3 +576,32 @@ func (c *client) StopAgent() error {
...
+		return false
+	}
+
+	realmOutputStr := strings.TrimSpace(string(realmCmdOutput))
+	if !strings.Contains(realmOutputStr, "realm-name") {
+		log.Errorf("couldn't not find realm name")
```
Comment (singholt) [inline ecs-init/docker/docker.go:602]: `log.Error`

File: ecs-init/docker/docker.go [hunk 28]
```diff
@@ -541,3 +576,32 @@ func (c *client) StopAgent() error {
...
+	}
+
+	realmCommmand := realmPath + " list"
+	realmCmdOutput, err := exec.Command("bash", "-c", realmCommmand).Output()
+	if err != nil {
+		log.Errorf("failed to read realm info")
+		return false
+	}
+
+	realmOutputStr := strings.TrimSpace(string(realmCmdOutput))
```
Comment (singholt) [inline ecs-init/docker/docker.go:600]: nit: do you need to do `TrimSpace` if you're doing a `string.Contains` on `realmOutputStr`?

File: ecs-init/docker/docker.go [hunk 29]
```diff
@@ -541,3 +576,32 @@ func (c *client) StopAgent() error {
 	}
 	return err
 }
+
+// isDomainJoined is used to validate if container instance is part of a valid active directory.
+func isDomainJoined() bool {
```
Comment (singholt) [inline ecs-init/docker/docker.go:581]: nit: can you add a unit test for this method? (you would need to mock `exec` methods)
example: https://github.com/aws/amazon-ecs-agent/blob/71a0f6765cc469913a0b09728196feb0d0d33e37/agent/taskresource/fsxwindowsfileserver/fsxwindowsfileserver_windows_test.go#L395

File: ecs-init/docker/docker.go [hunk 30]
```diff
@@ -440,6 +459,22 @@ func (c *client) getHostConfig(envVarsFromFiles map[string]string) *godocker.Hos
 	return createHostConfig(binds)
 }
 
+// getCredentialsFetcherSocketBind returns the corresponding bind for credentials fetcher socket.
+func getCredentialsFetcherSocketBind() (string, bool) {
+	CredentialsFetcherUnixSocketHostPath := config.CredentialsFetcherUnixSocket()
```
Comment (singholt) [inline ecs-init/docker/docker.go:464]: `config.HostCredentialsFetcherPath()` calls `CredentialsFetcherUnixSocket()` to get the unix socket host path right? can we use `config.HostCredentialsFetcherPath()` in all places in `docker.go` and `CredentialsFetcherUnixSocket` can be local to `config.go`?

File: ecs-init/docker/docker.go [hunk 31]
```diff
@@ -440,6 +459,22 @@ func (c *client) getHostConfig(envVarsFromFiles map[string]string) *godocker.Hos
...
+	CredentialsFetcherUnixSocketAgentPath := CredentialsFetcherUnixSocketHostPath
```
Comment (singholt) [inline ecs-init/docker/docker.go:465]: nit: do we need a new var `CredentialsFetcherUnixSocketAgentPath`?

could we just return `CredentialsFetcherUnixSocketHostPath + ":" + CredentialsFetcherUnixSocketHostPath`

File: ecs-init/docker/docker.go (hunk 30 above)
Comment (saikiranakula-amzn) [reply ecs-init/docker/docker.go:464]: config.go is in ecs-init and docker.go is in ecs-agent code base, that is not possible

File: ecs-init/docker/docker.go (hunk 31 above)
Comment (saikiranakula-amzn) [reply ecs-init/docker/docker.go:465]: just to make it visible about the bind mount

File: ecs-init/docker/docker.go (hunk 29 above)
Comment (saikiranakula-amzn) [reply ecs-init/docker/docker.go:584]: added tests

File: ecs-init/config/common.go [hunk 32]
```diff
@@ -212,6 +223,23 @@ func DockerUnixSocket() (string, bool) {
...
+func CredentialsFetcherUnixSocket() string {
+	if credentialsFetcherHost := os.Getenv(CredentialsFetcherHostEnvVar); strings.HasPrefix(credentialsFetcherHost, UnixSocketPrefix) {
+		return strings.TrimPrefix(credentialsFetcherHost, UnixSocketPrefix)
+	}
+
+	return DefaultCredentialsFetcherSocketPath
+}
+
+// HostCredentialsFetcherPath() returns the daemon socket location if it is available
+func HostCredentialsFetcherPath() (string, bool) {
```
[the remark this answers is not in the document]
Comment (saikiranakula-amzn) [reply ecs-init/config/common.go:236]: added unit test

File: agent/config/parse_linux.go [hunk 33]
```diff
@@ -18,10 +18,43 @@ package config
...
+			_, err := os.Stat(credentialsfetcherHostDir)
+			if err != nil {
+				if os.IsNotExist(err) {
+					seelog.Errorf("CREDENTIALS_FETCHER_HOST_DIR not found, was looking for %v", err)
```
[the remark this answers is not in the document]
Comment (saikiranakula-amzn) [reply agent/config/parse_linux.go]: rephrased

File: ecs-init/docker/docker.go (hunk 30 above)
Comment (singholt) [reply ecs-init/docker/docker.go:467]: i was referring to `ecs-init/config/`, this `CredentialsFetcherUnixSocket` resides in `ecs-init/config/common.go` isnt it?

File: ecs-init/docker/docker.go (hunk 31 above)
Comment (singholt) [reply ecs-init/docker/docker.go:468]: we're doing this today as well.

example: `binds = append(binds, gpu.GPUInfoDirPath+":"+gpu.GPUInfoDirPath)` above around L444

File: ecs-init/docker/docker_test.go [hunk 34]
```diff
@@ -985,3 +987,99 @@ func TestDefaultIsPathValid(t *testing.T) {
...
+	assert.False(t, out)
+}
+
+func TestExecHelperProcess(t *testing.T) {
+	if os.Getenv("GO_WANT_HELPER_PROCESS") != "1" {
+		return
+	}
+
+	mockRealmList := "contoso.com\n  type: kerberos\n  realm-name: CONTOSO.COM\n  domain-name: contoso.com\n  configured: kerberos-member\n  server-software: active-directory\n  client-software: sssd\n  required-package: oddjob\n  required-package: oddjob-mkhomedir\n  required-package: sssd\n  required-package: adcli\n  required-package: samba-common-tools\n  login-formats: %U@contoso.com\n  login-policy: allow-realm-logins"
+	fmt.Fprintf(os.Stdout, mockRealmList)
```
Comment (singholt) [inline ecs-init/docker/docker_test.go:1075]: nit: is this needed?

Comment (saikiranakula-amzn) [reply ecs-init/docker/docker_test.go:1075]: this how we pass the response back if I understood other examples right

File: ecs-init/docker/docker.go [hunk 35]
```diff
@@ -127,6 +127,9 @@ const (
 	execAgentLogRelativePath = "/exec"
 )
 
+var execCommand = exec.Command
```
Comment (singholt) [inline ecs-init/docker/docker.go:130]: nit: can you have all vars into one block?

File: ecs-init/docker/docker_test.go [hunk 36]
```diff
@@ -985,3 +987,99 @@ func TestDefaultIsPathValid(t *testing.T) {
...
+
+	out := isDomainJoined()
+	assert.True(t, out)
+}
+
+func TestInstanceDomainJoinedRealmNotFound(t *testing.T) {
+	execCommand = fakeExecCommand
+	defer func() { execCommand = exec.Command }()
+
+	defer func() { execLookPath = exec.LookPath }()
```
Comment (singholt) [inline ecs-init/docker/docker_test.go:1059]: super-nit: this can be defined after `execLookPath` for consistency with how you defined execCommand and its defer above.

Comment (singholt) [reply ecs-init/docker/docker_test.go:1059]: even better would be to have just 1 defer for both execCommand and execLookPath

File: ecs-init/docker/docker.go (hunk 30 above)
Comment (saikiranakula-amzn) [reply ecs-init/docker/docker.go]: refactored the code accordingly

File: ecs-init/docker/docker.go (hunk 31 above)
Comment (saikiranakula-amzn) [reply ecs-init/docker/docker.go]: refactored variables accordingly

File: ecs-init/docker/docker.go (hunk 35 above)
Comment (saikiranakula-amzn) [reply ecs-init/docker/docker.go]: moved to a common var

File: ecs-init/docker/docker_test.go (hunk 36 above)
Comment (saikiranakula-amzn) [reply ecs-init/docker/docker_test.go]: moved to common defer

**Review by singholt — DISMISSED** [review]
🚀

**Review by singholt — APPROVED** [review]

**Review by sparrc — APPROVED** [review]
